### PR TITLE
WIP: feat: new ast for typescript 3.8 

### LINKED ISSUE #1436: TypeScript 3.8 Syntax Support
3.8 requires new AST nodes and to support the new syntax. 

### Type Import (#1697) (_SUPPORTED_)
```ts
import type { SomeThing } from "./some-module.js";
export type { SomeThing };
```

### `export * as ns` Syntax (#1698) (_SUPPORTED_)
```TS
export * as utilities from "./utilities.js";
```

### ECMAScript Private Fields (_NOT YET SUPPORTED_ - see #3430)

AST: https://github.com/estree/estree/pull/222

**Important Note**
We currently "support" such that no tooling should fail on private fields.

However, the AST that is produced is not considered stable and ***will*** change in future _without a major release_. Any tooling looking to lint private fields should not rely upon the representation (or be prepared to handle a breakage).

**Why don't we have support yet?**

In short: we are awaiting agreement on the AST representation from the ESTree spec.  

For the 3.7 release we jumped the gun and eagerly implemented optional chaining using babel's AST representation, thinking that the ESTree spec, would just go with that. However ESTree is going with a different and incompatible representation entirely. This will cause our project (and users) a lot of pain as we will have to do a _major_, breaking change to the AST to align with what ESLint core will use.

To avoid going through the same pain again with private class fields, we're holding off on official support until ESTree has decided upon a representation.

There are a lot of projects, and a lot of volunteers that collaborate on the ESTree spec, so things take time; please be patient.

```ts
class Person {
    #name: string
    
    constructor(name: string) {
        this.#name = name;
    }

    greet() {
        console.log(`Hello, my name is ${this.#name}!`);
    }
}

let jeremy = new Person("Jeremy Bearimy");

jeremy.#name
```

### Read more:
https://devblogs.microsoft.com/typescript/announcing-typescript-3-8-beta/

### PR DESCRIPTION
**Pull request by armano2**
**DO NOT MERGE,**

This is branch contains so far only test scenarios for upcoming typescript 3.8.

---------------------

TODO:
- [ ] Private class properties: https://github.com/estree/estree/pull/180
- [ ] Private class method: https://github.com/estree/estree/pull/180
- [x] ~Private class properties in tsx~
- [x] Export type: https://github.com/babel/babel/issues/10981
- [x] Import type: https://github.com/babel/babel/issues/10981
- [ ] Export as ns from: https://github.com/estree/estree/pull/205

Implemented as draft (AST may change base on tickets above:
- [x] Private class properties
- [x] Private class method
- [x] Export type
  - `exportKind: 'type' | 'value'`
- [x] Import type
  - `importKind: 'type' | 'value'`
- [x] Export as ns from

Issues:
- [x] There is difference in range for `export * as ns` between babel and ts - https://github.com/estree/estree/pull/205
- [ ] createDiagnostic crashes on
      `import foo, {bar} from "foo";`
      `Cannot read property 'kind' of undefined`

https://github.com/prettier/prettier/issues/6608
https://github.com/prettier/prettier/issues/7263
https://github.com/benmosher/eslint-plugin-import/issues/645
fixes #1436

### CODE REVIEW

**Comment by armano2** [comment]
Private class properties in tsx are supported by typescript but so far they are not documented (i'm unsure if this is intended)

File: packages/typescript-estree/package.json [hunk 1]
```diff
@@ -48,9 +48,9 @@
     "tsutils": "^3.17.1"
   },
   "devDependencies": {
-    "@babel/code-frame": "7.5.5",
-    "@babel/parser": "7.7.7",
-    "@babel/types": "^7.7.4",
+    "@babel/code-frame": "7.8.3",
+    "@babel/parser": "7.8.3",
+    "@babel/types": "^7.8.3",
```
Comment (armano2) [inline packages/typescript-estree/package.json]: i should do this upgrade outside of this PR

File: packages/typescript-estree/src/convert.ts [hunk 2]
```diff
@@ -1563,9 +1579,10 @@ export class Converter {
           return this.createNode<TSESTree.ExportNamedDeclaration>(node, {
             type: AST_NODE_TYPES.ExportNamedDeclaration,
             source: this.convertChild(node.moduleSpecifier),
-            specifiers: node.exportClause.elements.map(el =>
-              this.convertChild(el),
-            ),
+            specifiers:
+              node.exportClause.kind === SyntaxKind.NamedExports
+                ? node.exportClause.elements.map(el => this.convertChild(el))
+                : [this.convertChild(node.exportClause.name)],
```
Comment (armano2) [inline packages/typescript-estree/src/convert.ts]: this is not valid we should return different node in this case, `* as ns` and its here just as temp workaround to make it compile

**Comment by nicolo-ribaudo** [comment]
>  There is difference in range for export * as ns between babel and ts

Which is the difference? Is it something that should be fixed in Babel?

**Comment by armano2** [comment]
@nicolo-ribaudo for code
```ts
export * as ns from 'x'
```
babel returns `ExportNamespaceSpecifier` node at char 7, typescript node `NamespaceExport` (262) is reported at 12 

babel:  `* as ns` - [astexplorer](https://astexplorer.net/#/gist/9f08ffb81890c0d4ff94018d1ef07b6e/6c576942df12b99af1341647fe9c084eef55fb52)
typescript: `ns` - [ts-ast-viewer](https://ts-ast-viewer.com/#code/KYDwDg9gTgLgBAKjgQwM5wHboGZQgWzgHIQig)

--------------

it looks like typescript returns same range for NamespaceExport and Identifier

**Comment by armano2** [comment]
@nicolo-ribaudo as think that is missing in babel is this weird syntax, i will fill bug report for this latter
```ts
class X {
	private #bar: string = ''
}
```

File: packages/typescript-estree/src/convert.ts [hunk 3]
```diff
@@ -1555,6 +1572,7 @@ export class Converter {
         return this.createNode<TSESTree.ImportDefaultSpecifier>(node, {
           type: AST_NODE_TYPES.ImportDefaultSpecifier,
           local: this.convertChild(node.name),
+          importKind: node.isTypeOnly ? 'type' : 'value',
```
Comment (kendallgassner) [inline packages/typescript-estree/src/convert.ts]: 👍 excited to add this to my eslint plugin!

**Comment by dsherret** [comment]
FYI that range difference was a bug that is now fixed: https://github.com/microsoft/TypeScript/issues/36762

**Comment by NickHeiner** [comment]
Is this PR actively being worked on? It looks like there haven't been any commits for a while.

**Comment by bradzacher** [comment]
There's a lot of work that needs to happen here.

For import/export type, they are TS specific features, so we are relatively free to implement them. However we need to be sure we're syncing up with babel so that our AST representations match. Both packages are part of one ecosystem, so we need to match up.

With the 3.7 release, we made the decision to jump the gun on the AST for both nullish coalescing and optional chaining, and used babel's representation. However it looks like the ESTree spec will go with a different representation, which ESLint will then build on top of. So in order to support the ecosystem, we'll have to migrate to that representation via a non-trival amount of work, and a major version bump.
This will cause a headache for any plugins using the current AST, so we want to avoid jumping the gun again.

So for private class fields/methods, and `export * as ns`, we taking it slow for now, to see what the official spec work is heading toward to minimise this disruption.

There is a lot of *volunteers* working on this across many projects, so it takes time as they find moments in their free time to work on it.

**Comment by NickHeiner** [comment]
Thanks for that context. I've updated my original comment to be clear that I don't feel entitled to anything. I appreciate the volunteers.

**Pull request closed without merging**